util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+                    defined[idx] = True
+
+        # Reshape this into 64bit blocks (this must be aligned at this point)
+        if len(data_bytes) % 8:
+            raise RuntimeError('Partitions must be 64bit aligned')
+
+        data_blocks = []
+        data_block_defined = []
+        for k, b in enumerate(data_bytes):
+            if (k % 8) == 0:
+                data_blocks.append(b)
+                data_block_defined.append(defined[k])
+            else:
+                data_blocks[k // 8] += (b << 8 * (k % 8))
+                # If any of the individual bytes are defined, the
+                # whole block is considered defined.
+                data_block_defined[k // 8] |= defined[k]
+
+        # Check if scrambling is needed
+        if self.config['partitions'][part_idx]['secret']:
+            part_name = self.config['partitions'][part_idx]['name']
+            key_sel = self.config['partitions'][part_idx]['key_sel']
+            log.info('> Scramble partition with key "{}"'.format(key_sel))
+
+            for key in self.config['scrambling']['keys']:
+                if key['name'] == key_sel:
+                    break
+            else:
+                raise RuntimeError(
+                    'Scrambling key cannot be found {}'.format(key_sel))
+
+            for k in range(len(data_blocks)):
+                if data_block_defined[k]:
+                    data_blocks[k] = _present_64bit_encrypt(
+                        data_blocks[k], key['value'])
+
+        # Check if digest calculation is needed
+        if self.config['partitions'][part_idx]['hw_digest']:
+            # Digest is stored in last block of a partition
+            if self.config['partitions'][part_idx].setdefault('lock', False):
+                log.info('> Lock partiton by computing digest')
+                # Digest constants at index 0 are used to compute the
+                # consistency digest
+                iv = self.config['scrambling']['digests'][0]['iv_value']
+                const = self.config['scrambling']['digests'][0]['cnst_value']
+                data_blocks[-1] = _present_64bit_digest(
+                    data_blocks[0:-1], iv, const)
+            else:
+                log.info('> Partition is not locked, hence assigning 0 digest')
+                data_blocks[-1] = 0
+
+        # Convert to a list of bytes to make final packing into
+        # OTP memory words independent of the cipher block size.
+        data = []
+        for block in data_blocks:
+            for k in range(8):
+                data.append((block >> (8 * k)) & 0xFF)
+
+        # Size sanity check
+        act_size = len(data)
+        exp_size = self.config['partitions'][part_idx]['size']
+        if exp_size < act_size:
+            raise RuntimeError(
+                'Stream out for partition {} failed.'.format(part_name))
+            raise RuntimeError('expected < actual size: {}B / {}B'.format(
+                exp_size, act_size))
```

Comment:
That won't work (because you'll never see the second error). Probably join together the two format calls? Also, it seems quite late to only figure this out here :-) Isn't `len(data)` always going to equal `part_size`, so we could have spotted this going wrong in the constructor? Maybe a nice error earlier plus an assertion here?

Reply:
yeah sorry this is a weird artifact from search replacing log.error with `RuntimeError` I have in the meantime fixed this and changed this to an assertion.